anywidget/_descriptor.py

```diff
@@ -0,0 +1,615 @@
+    """Check if an object has a psygnal.SignalGroup, and connect it to send_state.
+
+    Returns
+    -------
+    disconnect : Callable | None
+        A callable that disconnects the psygnal.SignalGroup from send_state, or None
+        if no psygnal.SignalGroup was found.
+    """
+    events = _get_psygnal_signal_group(obj)
+
+    if events is not None:
+
+        @events.connect
+        def _on_psygnal_event(event: psygnal.EmissionInfo):
+            send_state({event.signal.name})
+
+        def _disconnect():
+            events.disconnect(_on_psygnal_event)
+
+        return _disconnect
+    return None
+
+
+# ------------- Traitlets support --------------
+
+
+def _is_traitlets_object(obj: Any) -> TypeGuard[traitlets.HasTraits]:
+    """Return `True` if an object is an instance of traitlets.HasTraits."""
+    traitlets = sys.modules.get("traitlets")
+    return isinstance(obj, traitlets.HasTraits) if traitlets is not None else False
+
+
+# a tag that can be added to a traitlet to indicate that it should be synced
+# this apparently comes from ipywidgets, not traitlets
+_TRAITLETS_SYNC_FLAG = "sync"
+
+
+# TODO: decide about usage of "sync" being opt-in or opt-out
+# users of traitlets who *don't* use ipywidgets might be surprised when their
+# state isn't being synced without opting in.
+
+
+def _get_traitlets_state(obj: traitlets.HasTraits) -> dict:
+    """Get the state of a traitlets.HasTraits instance."""
+    return obj.trait_values(**{_TRAITLETS_SYNC_FLAG: True})
+
+
+def _connect_traitlets(obj: object, send_state: Callable) -> Callable | None:
+    """Check if an object has a psygnal.SignalGroup, and connect it to send_state.
+
+    Returns
+    -------
+    disconnect : Callable | None
+        A callable that disconnects the psygnal.SignalGroup from send_state, or None
+        if no psygnal.SignalGroup was found.
+    """
+    if not _is_traitlets_object(obj):
+        return None
+
+    obj_ref = weakref.ref(obj)
+
+    @obj.observe
+    def _on_trait_change(change: dict):
+        name = change["name"]
+        if obj_ref().trait_metadata(name, _TRAITLETS_SYNC_FLAG):
+            send_state({name})
+
+    def _disconnect():
+        obj = obj_ref()
+        if obj is not None:
+            obj.unobserve(_on_trait_change)
+
+    return _disconnect
+
+
+# ------------- Pydantic support --------------
+
+
+def _is_pydantic_model(obj: Any) -> TypeGuard[pydantic.BaseModel]:
+    """Return `True` if an object is an instance of pydantic.BaseModel."""
+    pydantic = sys.modules.get("pydantic")
+    return isinstance(obj, pydantic.BaseModel) if pydantic is not None else False
+
+
+def _get_pydantic_state(obj: pydantic.BaseModel) -> dict:
+    """Get the state of a pydantic BaseModel instance.
+
+    To take advantage of pydantic's support for custom encoders (with json_encoders)
+    we call obj.json() here, and then cast back to a dict (which is what the comm
+    expects.)
+    """
+    return json.loads(obj.json())
```

Comment:
Yeah really a fan of this

Reply:
Although kind of a bummer that we need deserialize the JSON into a dict. I guess this is something we could follow up with some special handling in the font-end:

```python
def _get_pydantic_state(obj: pydantic.BaseModel) -> dict:
    return { "_pydantic_json": obj.json() }
```

Reply:
yeah, and I guess we could also look at the Comm object itself.  since we're creating it now, we could look at that source and possible modify the final serialization

Reply:
let's defer this for a later PR